src/cis2/types.rs

```diff
@@ -0,0 +1,645 @@
+    type Err = ParseTokenIdVecError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = hex::decode(s)?;
+        if bytes.len() > 255 {
+            Err(ParseTokenIdVecError::TooManyBytes(bytes.len()))
+        } else {
+            Ok(TokenIdVec(bytes))
+        }
+    }
+}
+
+/// Display the token ID as a hex string.
+impl std::fmt::Display for TokenIdVec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        for b in &self.0 {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+/// Serialize the token ID according to CIS2 specification.
+impl Serial for TokenIdVec {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        serial_vector_no_length(&self.0, out)
+    }
+}
+
+/// Deserialize bytes to a Token ID according to CIS2 specification.
+impl Deserial for TokenIdVec {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let tokens_id_length = u8::deserial(source)?;
+        let bytes = deserial_vector_no_length(source, tokens_id_length.into())?;
+        Ok(TokenIdVec(bytes))
+    }
+}
+
+/// The different errors found in CIS2.
+#[derive(Debug, PartialEq, Eq, Error)]
+pub enum Cis2ErrorRejectReason {
+    /// Invalid token id (Error code: -42000001).
+    #[error("Invalid token ID")]
+    InvalidTokenId,
+    /// The balance of the token owner is insufficient for the transfer (Error
+    /// code: -42000002).
+    #[error("Insufficient funds for the transfer")]
+    InsufficientFunds,
+    /// Sender is unauthorized to call this function (Error code: -42000003).
+    #[error("Sender is unauthorized to call this function")]
+    Unauthorized,
+    /// Unknown error code for CIS2.
+    #[error("Non-CIS2 error code: {0}")]
+    Other(i32),
+}
+
+/// Convert Cis2Error into a reject with error code:
+/// - InvalidTokenId: -42000001
+/// - InsufficientFunds: -42000002
+/// - Unauthorized: -42000003
+impl From<i32> for Cis2ErrorRejectReason {
+    fn from(error_code: i32) -> Self {
+        match error_code {
+            -42000001 => Cis2ErrorRejectReason::InvalidTokenId,
+            -42000002 => Cis2ErrorRejectReason::InsufficientFunds,
+            -42000003 => Cis2ErrorRejectReason::Unauthorized,
+            other => Cis2ErrorRejectReason::Other(other),
+        }
+    }
+}
+
+/// Additional data which can be included for each transfer in the
+/// transfer parameter for the CIS2 contract function `transfer`.
+#[derive(Debug, Clone, From, AsRef)]
+pub struct AdditionalData {
+    pub data: Vec<u8>,
+}
+
+/// Serialization for the additional data, serialized as according to the CIS2
+/// specification.
+impl Serial for AdditionalData {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u16::try_from(self.data.len()).map_err(|_| W::Err::default())?;
```

Comment:
You are relyin on the length fitting into a u16 here, but this is not documented as part of the type. And the type has a `From` implementation which is not correct if the length must fit into u16.

It would be better to have `::new` and `::new_unchecked` methods instead, and the private field.

Reply:
> It would be better to have ::new and ::new_unchecked methods instead, and the private field.
I went through the types added this